flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/AlterTableSchemaUtil.java

```diff
@@ -0,0 +1,783 @@
+                            new SqlCallExpression(escapeExpressions.apply(validated))));
+        }
+
+        private SqlNode verifyWatermark(
+                AlterSchemaStrategy strategy, String rowtimeField, SqlNode watermarkStrategy) {
+            Map<String, RelDataType> nameToTypeMap = new HashMap<>();
+            nameToTypeMap.putAll(physicalFieldNamesToTypes);
+            nameToTypeMap.putAll(metadataFieldNamesToTypes);
+            nameToTypeMap.putAll(computedFieldNamesToTypes);
+            verifyRowtimeAttribute(strategy, rowtimeField, nameToTypeMap);
+
+            // this will validate and expand function identifiers
+            SqlNode validated =
+                    sqlValidator.validateParameterizedExpression(watermarkStrategy, nameToTypeMap);
+
+            String fullTypeString =
+                    sqlValidator.getValidatedNodeType(validated).getFullTypeString();
+            // ignore the nullability comparison
+            if (!fullTypeString.contains("TIMESTAMP(3)")) {
+                throw new ValidationException(
+                        String.format(
+                                "Could not %s watermark spec because the watermark spec %s produces the type '%s', but 'TIMESTAMP(3)' is required.",
+                                strategy == AlterSchemaStrategy.ADD ? "create" : "modify",
+                                escapeExpressions.apply(watermarkStrategy),
+                                fullTypeString));
+            }
+            return validated;
+        }
+
+        private void verifyWatermark() {
+            if (!watermarkSpecs.isEmpty()) {
+                Map<String, RelDataType> nameToTypeMap = new HashMap<>();
+                nameToTypeMap.putAll(physicalFieldNamesToTypes);
+                nameToTypeMap.putAll(metadataFieldNamesToTypes);
+                nameToTypeMap.putAll(computedFieldNamesToTypes);
+                for (Map.Entry<String, Schema.UnresolvedWatermarkSpec> entry :
+                        watermarkSpecs.entrySet()) {
+                    String columnName = entry.getKey();
+                    String watermarkStrategy =
+                            ((SqlCallExpression) entry.getValue().getWatermarkExpression())
+                                    .getSqlExpression();
+                    verifyRowtimeAttribute(AlterSchemaStrategy.MODIFY, columnName, nameToTypeMap);
+                    try {
+                        CalciteParser parser = new CalciteParser(parserConfig);
+                        SqlNode validated =
+                                sqlValidator.validateParameterizedExpression(
+                                        parser.parseExpression(watermarkStrategy), nameToTypeMap);
+
+                        String fullTypeString =
+                                sqlValidator.getValidatedNodeType(validated).getFullTypeString();
+                        // ignore the nullability comparison
+                        if (!fullTypeString.contains("TIMESTAMP(3)")) {
```

Comment:
`TIMESTAMP(3) NOT NULL` is not required because `TIMESTAMP(3)` is valid to generate a plan, and the watermark rowtime field might be a derived one. E.g. `TO_TIMESTAMP` always returns `TIMESTAMP(3)` instead of `TIMESTAMP(3) NOT NULL`.